ibis/expr/rewrites.py

```diff
@@ -0,0 +1,60 @@
+"""Some common rewrite functions to be shared between backends."""
+from __future__ import annotations
+
+import functools
+from collections.abc import Mapping
+
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+from ibis.common.patterns import pattern, replace
+from ibis.util import Namespace
+
+p = Namespace(pattern, module=ops)
+
+
+@replace(p.FillNa)
+def rewrite_fillna(_):
+    """Rewrite FillNa expressions to use more common operations."""
+    if isinstance(_.replacements, Mapping):
```

Comment:
Haven't taken a thorough look yet, just a quick look. One thing we may want to consider is to have more specific matchers in the rewrites to keep the replacement rules simpler and take their composition at the end.

In this scenario we clearly have a too generic `FillNa` operation (we should have a todo note about splitting it). Nonetheless we could match certain variants of the `FillNa` instances, according to the `FillNa.replacements` typehint we have two cases:

`UnionType[Value[dt.Numeric | dt.String], FrozenDict[str, Any]]`


```py
p.FillNa(replacements=p.Value)
```

```py
p.FillNa(replacements=Mapping) # or p.FillNa(replacements=FrozenDict)
```

we can also match on empty dictionary,  this should work since the dict value is turned into a more robuts dictionary matcher (though the empty case should be optimized to a `InstanceOf(Mapping)` pattern in a new `PatternMapping.__create__()`)

```py
p.FillNa(replacements={})  
```

So we *could* have 2 or three simpler replacement rules and then combine them with `|`. 

I'm not sure that should be the desired way to do it, after all we rewrite `FillNa` operations no matter what, not even sure whether would that result in more maintainable rules or not, but it worths discussing at least.

Reply:
> I'm not sure that should be the desired way to do it, after all we rewrite FillNa operations no matter what, not even sure whether would that result in more maintainable rules or not, but it worths discussing at least.

I agree it's worth discussing. Personally I like having all the code for this localized - it makes it clearer IMO to see how the specific operation (`fillna`) is simplified. Splitting it up into a few rewrite rules (or even splitting the op into 2 versions) would be a neutral to negative change in my mind. I'd expect the resulting code to be longer, and the loss of localization may make it harder to reason about.

Either way, this is a topic for a future PR.